src/merkle/index.rs

```diff
@@ -97,6 +104,21 @@ impl NodeIndex {
         self.depth == 0
     }
 
+    /// Return an iterator with the directions to take from root to arrive at the target index
+    pub fn directions(&self) -> Vec<Direction> {
+        (0..self.depth)
+            .rev()
+            .map(|bit_pos| {
+                let is_odd = (1u64 << bit_pos) & self.value == 0b1;
+                if is_odd {
+                    Direction::Right
+                } else {
+                    Direction::Left
+                }
+            })
+            .collect()
+    }
+
```

Comment:
If the purpose is to speed up the implementation, then we are good to go. We just need to remember in the short term to refactor this part and probably implement a `move_down` or iterator.

Reply:
I'd actually prefer to do this in a way which is similar to the bit iterator. I think if we are traversing from the root down, that's probably the best approach.

Reply:
I'm going the change, but we lose a little in readability, here we have the nice values of `Left` and `Right` instead of bits `true` and `false`. Would it be okay if I translate the bit iterator to a direction iterator?

Reply:
Not a blocker for me, but I'd rather stick to `bool` both for simplicity (not introduce new types) & performance

Reply:
I'm not sure if counting the number of types is a good metric of simplicity. On a large scale that implies in a god object design, in a small scale that would mean not using types like `Option`. God objects are notoriously hard to maintain because of lack of encapsulation, small types like `Option` are super useful to express intent and actually makes the code more readable, not using them is more of a loss than a win, that is why this pattern is used pervasively through out the stdlib and it is part of the [API guidelines](https://rust-lang.github.io/api-guidelines/type-safety.html#arguments-convey-meaning-through-types-not-bool-or-option-c-custom-type).

I understand performance is a concern, but I'm not sure if that important here. `enum` are basically integers and can easily be optimized by the compiler, and if it needs a hint we can also use `transparent` repr. But again, I'm not sure if it is warranted to worry about that at this scale.

Reply:
It's a Occam razor principle. If we have two solutions for a problem, and one is simpler & more performant (even if the difference is small), then we probably should stick to that option.

But as mentioned; not a blocker for me.

Reply:
On a side note about performance. I think it is a super important goal that we have to have in mind, but I also think we should always have that in perspective, an off-the-shelf CPU nowadays operates on ranges from 1Ghz to 4Ghz, which means 1 to 4 billion operations, even mobile CPUs are super scalar and should be able to retire a few instructions in parallel, which optimistically bumps these values by a multiple from 1 to 4 (depending on hardware ofcourse). And we also have something from 1 to 4 cpu cores. So we are talking on the ballpark of 1billion to 64billion operations per second. I think we should not worry about a few instructions unless it is in the hot path, and to know what is the hot path we need benchmarking.

With that said, I'm not saying that we should intentionally write code that performs badly, or does unnecessary work. I think it is fine to remove unnecessary allocations here and there since these stack up and can have lock contention. Algorithmic complexity should also be paid attention to since these do stack up quite quickly. But I'm saying that counting instructions should be done with the support of benchmarks, and that we should focus on this style of optimization when we have a known bottleneck on the hot path that can't be fixed with a new algorithm with a smaller complexity.